File: rap-extensions/bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/internal/lifecycle/WidgetUtil.java
/*******************************************************************************
 * Copyright (c) 2002, 2015 Innoopract Informationssysteme GmbH and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Innoopract Informationssysteme GmbH - initial API and implementation
 *    EclipseSource - ongoing development
 ******************************************************************************/
package org.eclipse.rap.rwt.internal.lifecycle;

import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.internal.util.ParamCheck;
import org.eclipse.swt.internal.widgets.WidgetTreeUtil;
import org.eclipse.swt.internal.widgets.WidgetTreeVisitor;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Widget;


public final class WidgetUtil {

  private WidgetUtil() {
    // prevent instantiation
  }

  public static RemoteAdapter getAdapter( Widget widget ) {
    RemoteAdapter adapter = widget.getAdapter( RemoteAdapter.class );
    if( adapter == null ) {
      throw new IllegalStateException( "Could not retrieve an instance of WidgetAdapter." );
    }
    return adapter;
  }

  public static String getId( Widget widget ) {
    return getAdapter( widget ).getId();
  }

  public static String getVariant( Widget widget ) {
    return ( String )widget.getData( RWT.CUSTOM_VARIANT );
  }

  public static <T extends Widget> WidgetLCA<T> getLCA( T widget ) {
    WidgetLCA<T> lca = widget.getAdapter( WidgetLCA.class );
    if( lca == null ) {
      throw new IllegalStateException( "Could not retrieve an instance of WidgetLCA." );
    }
    return lca;
  }

  public static Widget find( Composite root, final String id ) {
    final Widget[] result = { null };
    if( id != null ) {
      WidgetTreeUtil.accept( root, new WidgetTreeVisitor() {
        @Override
        public boolean visit( Widget widget ) {
          if( getId( widget ).equals( id ) ) {
            result[ 0 ] = widget;
          }
          return result[ 0 ] == null;
        }
      } );
    }
    return result[ 0 ];
  }

  public static void registerDataKeys( String... keys ) {
    ParamCheck.notNull( keys, "keys" );
    WidgetDataUtil.registerDataKeys( keys );
  }

}

File: rap-extensions/bundles/org.eclipse.rap.rwt/js/rwt/widgets/util/WidgetUtil.js
/*******************************************************************************
 * Copyright (c) 2007, 2015 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Innoopract Informationssysteme GmbH - initial API and implementation
 *    EclipseSource - ongoing development
 ******************************************************************************/

/**
 * This class contains static helper functions for widgets.
 */
rwt.qx.Class.define( "rwt.widgets.util.WidgetUtil", {

  statics : {

    getControl : function( widget ) {
      var widgetManager = rwt.remote.WidgetManager.getInstance();
      var result = widget;
      while( result != null && !widgetManager.isControl( result ) ) {
        result = result.getParent ? result.getParent() : null;
      }
      return result;
    },

    getShell : function( widget ) {
      var result = widget;
      while( result != null && result.classname !== "rwt.widgets.Shell" ) {
        result = result.getParent();
      }
      return result;
    },

    getChildIds : function( widget ) {
      return widget.getUserData( "rwt_Children" );
    },

    isVisible : function( widget ) {
      var result = true;
      var current = widget;
      while( current && result ) {
        result = current.getVisibility();
        current = current.getParent();
      }
      return result;
    },

    getGC : function( widget ) {
      var gc = widget.getUserData( "rwt.widgets.GC" );
      if( gc == null ) {
        gc = new rwt.widgets.GC( widget );
        widget.setUserData( "rwt.widgets.GC", gc );
      }
      return gc;
    },

    callWithElement : function( widget, callback ) {
      if( widget.getElement() ) {
        callback( widget.getElement() );
      } else {
        var listener = function listener(){
          callback( widget.getElement() );
          widget.removeEventListener( "create", listener );
        };
        widget.addEventListener( "create", listener );
      }
    },

    isWidget : function( target ) {
      return    ( ( target.classname || "" ).indexOf( "rwt.widgets" ) === 0 )
        || ( target instanceof rwt.widgets.base.Widget );
    },

    /**
     * Can be used simulate mouseEvents on the qooxdoo event-layer.
     * Manager and handler that are usually notified by
     * rwt.event.EventHandler will not receive the event.
     */
    _fakeMouseEvent : function( originalTarget, type ) {
      if( !( originalTarget instanceof rwt.widgets.base.Widget ) ) {
        originalTarget = rwt.event.EventHandlerUtil.getTargetObject( originalTarget );
        originalTarget = originalTarget || rwt.widgets.base.ClientDocument.getInstance();
      }
      if( originalTarget.getEnabled() ) {
        var domTarget = originalTarget._getTargetNode();
        var EventHandlerUtil = rwt.event.EventHandlerUtil;
        var target = EventHandlerUtil.getTargetObject( null, originalTarget, true );
        var domEvent = {
          "type" : type,
          "target" : domTarget,
          "button" : 0,
          "wheelData" : 0,
          "detail" : 0,
          "pageX" : 0,
          "pageY" : 0,
          "clientX" : 0,
          "clientY" : 0,
          "screenX" : 0,
          "screenY" : 0,
          "shiftKey" : false,
          "ctrlKey" : false,
          "altKey" : false,
          "metaKey" : false,
          "preventDefault" : function(){}
        };
        var event = new rwt.event.MouseEvent( type,
                                              domEvent,
                                              domTarget,
                                              target,
                                              originalTarget,
                                              null );
        target.dispatchEvent( event );
      }
    }

  }
});
